akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -60,10 +60,23 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
           Idle
         }
     }
+
   def handleStreamEvent(e: StreamFrameEvent): Unit = {
+    // only validate maxConcurrentStreams compliance if we'd increment the number
+    // of streams in one of the open states. This way, when the value of SETTINGS_MAX_CONCURRENT_STREAMS
+    // decreases, in-flight streams can still finish and close.
+    e match {
+      case _: ParsedHeadersFrame =>
+        checkMaxConcurrentStreamsCompliance(incomingStreams.size)
+      case _: PushPromiseFrame =>
+        checkMaxConcurrentStreamsCompliance(incomingStreams.size)
+      case _ =>
+    }
```

Comment:
> Should we check this after updating the state? When updating the state might increate the number of incoming streams?

Interesting. Let me have a look.
Hmm, it's a thin line. 

So, this check is only run when receiving a HEADERS or a PUSH_PROMISE frames. Should we process a HEADERS/PUSH_PROMISE which won't even create a stream? 
I mean, HEADERS/PUSH_PROMISE will generally create a stream. But there are cases in which a HEADERS/PUSH_PROMISE immediately transitions to `Closed` (when END_STREAM is set).

I agre that moving the check to _after_ `updateState()` sounds harmless but I'm not sure I see it being more correct.

What am I missing?

Reply:
I think it becomes even more relevant when using RST_STREAM instead of PROTOCOL_ERROR. In that case, we could move the logic to `expectIncomingStream`, use `resetStream` to refuse the stream right away and set the state to CLOSED. WDYT?